qemu/tests/chardev_pci_serial_login.py

```diff
@@ -0,0 +1,51 @@
+from virttest import error_context
+from virttest import remote
+from virttest import qemu_monitor
+from virttest import env_process
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Verify the login function of pci-serial (RHEL and x86 only):
+    1) Start guest with pci-serial with backend unix_socket
+    2) nc host/path for unix or tcp backend.
+    3) login to guest and move files
+    4) Hot-unplug chardev which is in use, should fail.
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    serial_id = params.objects('serials')[-1]
+    prompt = params.get("shell_prompt")
+    params['start_vm'] = 'yes'
+    for backend in ['tcp_socket', 'unix_socket']:
+        params['chardev_backend_%s' % serial_id] = backend
+        env_process.preprocess_vm(test, params, env, params['main_vm'])
+        vm = env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        serial_device = vm.devices.get(serial_id)
+        chardev_qid = serial_device.get_param("chardev")
+        chardev_device = vm.devices.get_by_qid(chardev_qid)[0]
+        if backend == 'unix_socket':
+            session = vm.wait_for_serial_login()
+        elif backend == 'tcp_socket':
+            session = remote.remote_login(client='nc',
+                                          host=chardev_device.params['host'],
+                                          port=chardev_device.params['port'],
+                                          username='root',
+                                          password='kvmautotest',
+                                          prompt=prompt,
+                                          timeout=240)
+        session.cmd('touch file.txt')
+        session.cmd('mkdir -p tmp')
+        session.cmd('command cp file.txt ./tmp/test.txt')
+        try:
+            vm.devices.simple_unplug(chardev_device, vm.monitor)
+        except qemu_monitor.QMPCmdError as e:
+            if 'is busy' not in e.data['desc']:
+                test.fail(e.data['desc'])
+        else:
+            test.fail("Hot-unplug should fail.")
+        session.close()
```

Comment:
shutdown vm is must for the first round, otherwise it will effect the vm for the second time.

Reply:
thanks and update, besides, I add the pty and file backend after confirming with libvirt. please help to review ~